src/Core/Form/IdentifiableObject/Builder/FormBuilder.php

```diff
@@ -110,28 +133,45 @@ public function getForm(array $data = [], array $options = [])
      */
     public function getFormFor($id, array $data = [], array $options = [])
     {
-        $data = array_merge($this->dataProvider->getData($id), $data);
+        // Fetch initial data
+        $initialData = $this->dataProvider->getData($id);
+        // Fetch initial options
+        $initialOptions = $options;
+
+        $this->hookDispatcher->dispatchWithParameters(
+            'action' . Container::camelize($this->registry->getType($this->formType)->getBlockPrefix()) . 'FormDataProviderData',
+            [
+                'data' => &$initialData,
+                'id' => $id,
+                'options' => &$initialOptions,
+            ]
+        );
+
+        // Merge data in initial data
+        $data = array_merge($initialData, $data);
+        // Merge initial options in options
+        $options = $initialOptions;
         if (null !== $this->optionsProvider) {
             $options = array_merge($this->optionsProvider->getOptions($id, $data), $options);
         }
 
         return $this->buildForm(
             $this->formType,
             $data,
-            $id,
-            $options
+            $options,
+            $id
         );
     }
 
     /**
      * @param string $formType
      * @param array $data
-     * @param int|null $id
      * @param array $options
+     * @param int|null $id
      *
      * @return FormInterface
      */
-    private function buildForm($formType, $data, $id = null, array $options = [])
+    private function buildForm($formType, $data, array $options = [], $id = null)
```

Comment:
Not that it changes much anyway but why switch these two parameters?

Reply:
`$options` is always used but not `$id` so I reorder for use.